core/modules/shrscan/Director.h

```diff
@@ -0,0 +1,315 @@
+//! @return true:  The query was cancelled or was scheduled to be cancelled.
+//-----------------------------------------------------------------------------
+
+virtual bool  cancelQuery(Query const* query, bool wait=false) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Cancel all shared scan queries.
+//!
+//! @param  wait    - When true, does not return until all queries have been
+//!                   actually cancelled (i.e. endQuery() method called).
+//!
+//! This method effectively calls cancelQuery() on each query.
+//-----------------------------------------------------------------------------
+
+virtual void  cancelAll(bool wait=false) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Obtain statistics about this shared scan director.
+//!
+//! @param  stats   - Where information is to be returned (see struct below).
+//!
+//! @return The number of queries actually running.
+//-----------------------------------------------------------------------------
+
+struct Statistics {
+    unsigned long long bytesLockMax;//< Maximum number of bytes to lock
+    unsigned long long bytesLocked; //< Actual  number of bytes locked in memory
+    unsigned long long msRunTotal;  //< Total milliseconds spent running
+    unsigned long long msIdleTotal; //< Total milliseconds spent idling
+    unsigned int       numChunks;   //< The number of chunks being handled
+    unsigned int       maxCanRun;   //< Maximum number of parallel queries
+    unsigned int       numQueued;   //< Number waiting to run
+    unsigned int       numRunning;  //< Number actually running
+    unsigned int       numSuspend;  //< Number suspended
+    unsigned int       numErrors;   //< Number of queries that had errors
+    unsigned int       numCancelled;//< Number of queries that were cancelled
+};
+
+virtual void  getStatistics(Statistics &stats) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Obtain query status.
+//!
+//! @param  query   - Pointer to a Query object that describes the query.
+//! @param  info    - Where information is to be returned (see struct below).
+//!
+//! @return false: The query was not found, status struct not touched.
+//! @return true:  The query information returned in status.
+//-----------------------------------------------------------------------------
+
+struct Status {
+    unsigned int   secInQ;    //< Seconds      waiting to run (wall clock)
+    unsigned int   msRunning; //< Milliseconds running so far (wall clock)
+    unsigned int   msIdling;  //< Milliseconds in idle state  (wall clock)
+    unsigned short pctChunks; //< Percentage of chunks processed times 100
+    bool           isRunning; //< When true is actually running
+    bool           isSuspend; //< When true is currently suspended
+};
+
+virtual bool  getStatus(Query const* query, Status &status) = 0;
```

Comment:
OK, we have aq conflict here. I don't want to throw an exception on 
something that is not exceptional (see previous mail file). I do need to 
tell you if I returned something. So, I am disposed to leave this as it 
is. It's the only reasonable compromise without contorting the code.

Andy

On Wed, 18 Nov 2015, Andy wrote:

>> +//! @param  info    - Where information is to be returned (see struct below).
>> +//!
>> +//! @return false: The query was not found, status struct not touched.
>> +//! @return true:  The query information returned in status.
>> +//-----------------------------------------------------------------------------
>> +
>> +struct Status {
>> +    unsigned int   secInQ;    //< Seconds      waiting to run (wall clock)
>> +    unsigned int   msRunning; //< Milliseconds running so far (wall clock)
>> +    unsigned int   msIdling;  //< Milliseconds in idle state  (wall clock)
>> +    unsigned short pctChunks; //< Percentage of chunks processed times 100
>> +    bool           isRunning; //< When true is actually running
>> +    bool           isSuspend; //< When true is currently suspended
>> +};
>> +
>> +virtual bool  getStatus(Query const* query, Status &status) = 0;
>
> Throw exception if query is not found? And return Status by value.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/lsst/qserv/pull/156/files#r45246323